columnflow/production/normalization.py

```diff
@@ -59,24 +63,85 @@ def get_stitching_datasets(self: Producer) -> list[od.Dataset]:
     return list(stitching_datasets)
 
 
-def get_br_from_inclusive_dataset(stats: dict) -> dict:
+def get_br_from_inclusive_dataset(self: Producer, inclusive_dataset: od.Dataset, stats: dict) -> dict[int, float]:
     """
     Helper function to compute the branching ratios from the inclusive sample.
     """
-    # get the sum of weights inclusive and per process
-    sum_mc_weight = stats["sum_mc_weight"]
-    sum_mc_weight_per_process = stats["sum_mc_weight_per_process"]
+    # define a helper variables and mapping between process ids and dataset names
+    proc_ds_map = {
+        d.processes.get_first().id: d
+        for d in self.config_inst.datasets
+        if d.name in stats.keys()
+    }
+    inclusive_proc = inclusive_dataset.processes.get_first()
+    N = lambda x: sn.Number(x, np.sqrt(x))  # alias for Number with counting error
+
+    # create a dictionary of branching ratios based on all gen weight sums
+    br: dict[int, dict[int, sn.Number]] = defaultdict(dict)
+    for proc, _, child_procs in inclusive_dataset.walk_processes():
+        if proc.id not in proc_ds_map or proc.is_leaf_process:
+            continue
+
+        # get the mc weights for the "mother" dataset and add an entry for the process
+        sum_mc_weight = stats[proc_ds_map[proc.id].name]["sum_mc_weight"]
+        sum_mc_weight_per_process = stats[proc_ds_map[proc.id].name]["sum_mc_weight_per_process"]
+        # use the number of events to compute the error on the branching ratio
+        num_events = stats[proc_ds_map[proc.id].name]["num_events"]
+        num_events_per_process = stats[proc_ds_map[proc.id].name]["num_events_per_process"]
+
+        # compute the branching ratios for the children wrt the mother process
+        for child_proc in child_procs:
+            # skip processes that are not covered by any dataset or irrelevant for the used dataset
+            # (identified as leaf processes that have no occurrences in the stats)
+            # (or as non-leaf processes that are not in the stitching datasets)
+            is_leaf = child_proc.is_leaf_process
+            if (
+                (is_leaf and str(child_proc.id) not in sum_mc_weight_per_process) or
+                (not is_leaf and child_proc.id not in proc_ds_map)
+            ):
+                continue
 
-    if not sum_mc_weight == sum(sum_mc_weight_per_process.values()):
-        raise Exception(
-            "sum of event weights does not match the sum of event weights per process",
-        )
+            proc_ids = [child_proc.id] if is_leaf else [p.id for p in child_proc.get_leaf_processes()]
+            # compute the uncertainty on the branching ratio using number of events
+            _br = N(sum(num_events_per_process.get(str(proc_id), 0) for proc_id in proc_ids)) / N(num_events)
+            # uncertainty is independent of the mc weights, so we can use the same relative uncertainty
```

Comment:
I am rather certain that this statement is wrong. We can probably make this assumption because most of the event weights are of similar order of magnitude, but if we had a sample where some weights are significantly larger than others, then these weight will dominate our uncertainty (extreme case: there is one weight much larger than all other weights combined --> in that case, the uncertainty will be completely dominated by that single event)